File: android-app/src/main/kotlin/com/jworks/kanjijourney/android/ui/theme/GlassTheme.kt
package com.jworks.kanjijourney.android.ui.theme

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.asPaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.navigationBars
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.jworks.kanjijourney.android.R

// ── Brand colors (per-app, used sparingly: borders, icon tints, glows) ──

object GlassBrand {
    val KanjiJourney = Color(0xFFFF6B35)   // warm orange
    val KanjiLens = Color(0xFF0D9488)    // teal
    val EigoQuest = Color(0xFF4A90D9)    // blue
    val EigoLens = Color(0xFF4F46E5)     // indigo

    // Current app brand — change this per app module
    val current = KanjiJourney
}

// ── Typography ──────────────────────────────────────────────────────

val DmSansFamily = FontFamily(
    Font(R.font.dm_sans_light, FontWeight.Light),
    Font(R.font.dm_sans_regular, FontWeight.Normal),
    Font(R.font.dm_sans_medium, FontWeight.Medium),
    Font(R.font.dm_sans_bold, FontWeight.Bold),
)

// ── Glass color palette ────────────────────────────────────────────

val GlassBackground = Color(0xFF050508)
val GlassSurfaceLight = Color(0xFF12121E)
val GlassSurfaceDark = Color(0xFF08080F)
val GlassBorder = GlassBrand.current.copy(alpha = 0.28f)
val GlassBorderHover = GlassBrand.current.copy(alpha = 0.45f)
val GlassTextPrimary = Color.White
val GlassTextSecondary = Color.White.copy(alpha = 0.65f)
val GlassTextMuted = Color.White.copy(alpha = 0.4f)

val GlassCardGradient: Brush
    get() = Brush.linearGradient(
        colors = listOf(
            Color.White.copy(alpha = 0.10f),
            Color.White.copy(alpha = 0.04f)
        )
    )

val GlassCardDisabledGradient: Brush
    get() = Brush.linearGradient(
        colors = listOf(
            Color.White.copy(alpha = 0.05f),
            Color.White.copy(alpha = 0.02f)
        )
    )

// ── Reusable glass composables ─────────────────────────────────────

@Composable
fun GlassCard(
    modifier: Modifier = Modifier,
    shape: RoundedCornerShape = RoundedCornerShape(12.dp),
    borderColor: Color = GlassBorder,
    borderWidth: Dp = 1.dp,
    content: @Composable BoxScope.() -> Unit
) {
    Box(
        modifier = modifier
            .clip(shape)
            .background(GlassCardGradient, shape)
            .border(borderWidth, borderColor, shape),
        content = content
    )
}

@Composable
fun GlassChip(
    selected: Boolean,
    selectedColor: Color = Color(0xFFFF6B35),
    modifier: Modifier = Modifier,
    content: @Composable BoxScope.() -> Unit
) {
    val shape = RoundedCornerShape(6.dp)
    val bg = if (selected)
        selectedColor.copy(alpha = 0.30f)
    else
        Color.White.copy(alpha = 0.06f)
    val border = if (selected)
        selectedColor.copy(alpha = 0.40f)
    else
        Color.White.copy(alpha = 0.10f)

    Box(
        modifier = modifier
            .clip(shape)
            .background(bg, shape)
            .border(1.dp, border, shape)
            .padding(horizontal = 10.dp, vertical = 5.dp),
        content = content
    )
}

@Composable
fun GlassTopBar(
    title: @Composable () -> Unit,
    actions: @Composable RowScope.() -> Unit = {},
    navigationIcon: @Composable () -> Unit = {}
) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .background(
                Brush.verticalGradient(
                    colors = listOf(
                        Color.White.copy(alpha = 0.10f),
                        Color.White.copy(alpha = 0.05f)
                    )
                )
            )
            .drawBehind {
                // Orange warmth at the top edge
                drawRect(
                    brush = Brush.verticalGradient(
                        colors = listOf(
                            GlassBrand.current.copy(alpha = 0.14f),
                            Color.Transparent
                        )
                    )
                )
                // Bottom border — orange-tinted divider
                drawLine(
                    color = GlassBrand.current.copy(alpha = 0.40f),
                    start = Offset(0f, size.height),
                    end = Offset(size.width, size.height),
                    strokeWidth = 1.dp.toPx()
                )
            }
            .padding(horizontal = 4.dp)
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp)
                .padding(horizontal = 4.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            navigationIcon()
            Box(
                modifier = Modifier.weight(1f),
                contentAlignment = Alignment.CenterStart
            ) {
                title()
            }
            Row(content = actions)
        }
    }
}

@Composable
fun GlassNavigationBar(
    modifier: Modifier = Modifier,
    content: @Composable RowScope.() -> Unit
) {
    val navBarPadding = WindowInsets.navigationBars.asPaddingValues()
    Box(
        modifier = modifier
            .fillMaxWidth()
            .background(
                Brush.verticalGradient(
                    colors = listOf(
                        Color.White.copy(alpha = 0.05f),
                        Color.White.copy(alpha = 0.10f)
                    )
                )
            )
            .drawBehind {
                // Subtle orange glow at the top edge
                drawRect(
                    brush = Brush.verticalGradient(
                        colors = listOf(
                            GlassBrand.current.copy(alpha = 0.06f),
                            Color.Transparent
                        )
                    )
                )
                // Top divider line — orange-tinted
                drawLine(
                    color = GlassBrand.current.copy(alpha = 0.25f),
                    start = Offset(0f, 0f),
                    end = Offset(size.width, 0f),
                    strokeWidth = 1.dp.toPx()
                )
            }
            .padding(bottom = navBarPadding.calculateBottomPadding())
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(64.dp)
                .selectableGroup()
                .padding(horizontal = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            content = content
        )
    }
}

// ── Section background for visual rhythm ────────────────────────────

/**
 * Alternating section backgrounds to break visual monotony.
 * Use [variant] = true for darker sections, false for lighter tinted sections.
 */
@Composable
fun GlassSection(
    modifier: Modifier = Modifier,
    variant: Boolean = false,
    content: @Composable BoxScope.() -> Unit
) {
    val bg = if (variant) GlassSurfaceDark else GlassSurfaceLight
    Box(
        modifier = modifier
            .fillMaxWidth()
            .drawBehind {
                drawRect(color = bg)
                // Top gradient separator line
                drawRect(
                    brush = Brush.horizontalGradient(
                        colors = listOf(
                            Color.Transparent,
                            GlassBrand.current.copy(alpha = 0.15f),
                            Color.Transparent
                        )
                    ),
                    size = androidx.compose.ui.geometry.Size(size.width, 1.dp.toPx())
                )
            },
        content = content
    )
}

// ── Utility: brand-tinted border color ──────────────────────────────

fun glassBorderColor(color: Color = GlassBrand.current, alpha: Float = 0.28f): Color =
    color.copy(alpha = alpha)

File: ios-app/KanjiJourney/Theme/GlassTheme.swift
import SwiftUI

// MARK: - Brand colors (per-app, used sparingly: borders, icon tints, glows)

enum GlassBrand {
    static let kanjiJourney = Color(hex: 0xFF6B35)   // warm orange
    static let kanjiLens    = Color(hex: 0x0D9488)   // teal
    static let eigoQuest    = Color(hex: 0x4A90D9)   // blue
    static let eigoLens     = Color(hex: 0x4F46E5)   // indigo

    /// Current app brand color
    static let current = kanjiJourney
}

// MARK: - Glass color palette

enum GlassColors {
    static let background    = Color(hex: 0x050508)
    static let surfaceLight  = Color(hex: 0x12121E)
    static let surfaceDark   = Color(hex: 0x08080F)

    static let border        = GlassBrand.current.opacity(0.28)
    static let borderHover   = GlassBrand.current.opacity(0.45)

    static let textPrimary   = Color.white
    static let textSecondary = Color.white.opacity(0.65)
    static let textMuted     = Color.white.opacity(0.4)

    /// Linear gradient for glass card backgrounds
    static let cardGradient = LinearGradient(
        colors: [Color.white.opacity(0.10), Color.white.opacity(0.04)],
        startPoint: .top,
        endPoint: .bottom
    )

    /// Dimmed gradient for disabled/coming-soon cards
    static let cardDisabledGradient = LinearGradient(
        colors: [Color.white.opacity(0.05), Color.white.opacity(0.02)],
        startPoint: .top,
        endPoint: .bottom
    )
}

// MARK: - Glass typography (DM Sans)

enum GlassTypography {
    // Register DM Sans fonts — call once at app startup
    static func registerFonts() {
        let fontNames = ["dm_sans_light", "dm_sans_regular", "dm_sans_medium", "dm_sans_bold"]
        for name in fontNames {
            guard let url = Bundle.main.url(forResource: name, withExtension: "ttf") else { continue }
            CTFontManagerRegisterFontsForURL(url as CFURL, .process, nil)
        }
    }

    private static func dmSans(_ size: CGFloat, weight: Font.Weight) -> Font {
        let name: String
        switch weight {
        case .light:   name = "DMSans-Light"
        case .medium:  name = "DMSans-Medium"
        case .bold:    name = "DMSans-Bold"
        default:       name = "DMSans-Regular"
        }
        // Fallback to system if DM Sans not bundled
        if let _ = UIFont(name: name, size: size) {
            return .custom(name, size: size)
        }
        return .system(size: size, weight: weight)
    }

    // Display
    static let displayLarge  = dmSans(57, weight: .light)
    static let displayMedium = dmSans(45, weight: .light)
    static let displaySmall  = dmSans(36, weight: .light)

    // Headline
    static let headlineLarge  = dmSans(32, weight: .regular)
    static let headlineMedium = dmSans(28, weight: .medium)
    static let headlineSmall  = dmSans(24, weight: .medium)

    // Title
    static let titleLarge  = dmSans(22, weight: .medium)
    static let titleMedium = dmSans(16, weight: .medium)
    static let titleSmall  = dmSans(14, weight: .medium)

    // Body
    static let bodyLarge  = dmSans(16, weight: .regular)
    static let bodyMedium = dmSans(14, weight: .regular)
    static let bodySmall  = dmSans(12, weight: .regular)

    // Label
    static let labelLarge  = dmSans(14, weight: .medium)
    static let labelMedium = dmSans(12, weight: .medium)
    static let labelSmall  = dmSans(11, weight: .medium)
}

// MARK: - Reusable glass components

/// Frosted-glass card with gradient background and brand-tinted border.
struct GlassCard<Content: View>: View {
    var cornerRadius: CGFloat = 12
    var borderColor: Color = GlassColors.border
    var borderWidth: CGFloat = 1
    let content: () -> Content

    init(
        cornerRadius: CGFloat = 12,
        borderColor: Color = GlassColors.border,
        borderWidth: CGFloat = 1,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.cornerRadius = cornerRadius
        self.borderColor = borderColor
        self.borderWidth = borderWidth
        self.content = content
    }

    var body: some View {
        content()
            .background(GlassColors.cardGradient)
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius))
            .overlay(
                RoundedRectangle(cornerRadius: cornerRadius)
                    .stroke(borderColor, lineWidth: borderWidth)
            )
    }
}

/// Selectable chip with glass styling.
struct GlassChip<Content: View>: View {
    let selected: Bool
    var selectedColor: Color = GlassBrand.current
    let content: () -> Content

    init(
        selected: Bool,
        selectedColor: Color = GlassBrand.current,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.selected = selected
        self.selectedColor = selectedColor
        self.content = content
    }

    var body: some View {
        let bg = selected ? selectedColor.opacity(0.30) : Color.white.opacity(0.06)
        let border = selected ? selectedColor.opacity(0.40) : Color.white.opacity(0.10)

        content()
            .padding(.horizontal, 10)
            .padding(.vertical, 5)
            .background(bg)
            .clipShape(RoundedRectangle(cornerRadius: 6))
            .overlay(
                RoundedRectangle(cornerRadius: 6)
                    .stroke(border, lineWidth: 1)
            )
    }
}

/// Glass-styled top bar with gradient background and brand glow.
struct GlassTopBar<Title: View, Actions: View, NavIcon: View>: View {
    let title: () -> Title
    let actions: () -> Actions
    let navigationIcon: () -> NavIcon

    init(
        @ViewBuilder title: @escaping () -> Title,
        @ViewBuilder actions: @escaping () -> Actions = { EmptyView() },
        @ViewBuilder navigationIcon: @escaping () -> NavIcon = { EmptyView() }
    ) {
        self.title = title
        self.actions = actions
        self.navigationIcon = navigationIcon
    }

    var body: some View {
        HStack(spacing: 4) {
            navigationIcon()
            title()
                .frame(maxWidth: .infinity, alignment: .leading)
            actions()
        }
        .frame(height: 56)
        .padding(.horizontal, 8)
        .background(
            LinearGradient(
                colors: [Color.white.opacity(0.10), Color.white.opacity(0.05)],
                startPoint: .top,
                endPoint: .bottom
            )
        )
        .overlay(alignment: .top) {
            // Orange warmth at the top edge
            LinearGradient(
                colors: [GlassBrand.current.opacity(0.14), .clear],
                startPoint: .top,
                endPoint: .bottom
            )
            .frame(height: 20)
        }
        .overlay(alignment: .bottom) {
            // Bottom divider
            Rectangle()
                .fill(GlassBrand.current.opacity(0.40))
                .frame(height: 1)
        }
    }
}

/// Glass-styled bottom navigation bar.
struct GlassNavigationBar<Content: View>: View {
    let content: () -> Content

    init(@ViewBuilder content: @escaping () -> Content) {
        self.content = content
    }

    var body: some View {
        VStack(spacing: 0) {
            // Top divider
            Rectangle()
                .fill(GlassBrand.current.opacity(0.25))
                .frame(height: 1)

            HStack {
                content()
            }
            .frame(height: 64)
            .frame(maxWidth: .infinity)
            .padding(.horizontal, 8)
        }
        .background(
            LinearGradient(
                colors: [Color.white.opacity(0.05), Color.white.opacity(0.10)],
                startPoint: .top,
                endPoint: .bottom
            )
        )
        .overlay(alignment: .top) {
            // Subtle orange glow
            LinearGradient(
                colors: [GlassBrand.current.opacity(0.06), .clear],
                startPoint: .top,
                endPoint: .bottom
            )
            .frame(height: 12)
        }
    }
}

/// Alternating section backgrounds for visual rhythm.
struct GlassSection<Content: View>: View {
    var variant: Bool = false
    let content: () -> Content

    init(variant: Bool = false, @ViewBuilder content: @escaping () -> Content) {
        self.variant = variant
        self.content = content
    }

    var body: some View {
        content()
            .frame(maxWidth: .infinity)
            .background(variant ? GlassColors.surfaceDark : GlassColors.surfaceLight)
            .overlay(alignment: .top) {
                // Decorative top gradient separator
                LinearGradient(
                    colors: [.clear, GlassBrand.current.opacity(0.15), .clear],
                    startPoint: .leading,
                    endPoint: .trailing
                )
                .frame(height: 1)
            }
    }
}

// MARK: - Utility

/// Brand-tinted border color with adjustable opacity.
func glassBorderColor(_ color: Color = GlassBrand.current, alpha: Double = 0.28) -> Color {
    color.opacity(alpha)
}
